### PR TITLE
Improve lineSpit performance.

### PR DESCRIPTION
**Pull request by vdukhovni**
- Eliminate `counter` from `loop1` it is always 0.

- Eliminate use of `B.count` in loop2, it almost always performs
  worse, especially when called to just return one line at a time,
  but keep the newline (in contrast to `lines` which drops them).
  In realistic use-cases the new code is ~30x faster!

  Instead count up to the desired number of lines or chunk end by
  using a copy of `B.elemIndex`, returning either (Right) the
  position of the target boundary, or (Left) the total number of
  lines read that still failed to reach the boundary.  This uses
  `B.memchr()`, which is generally faster than `B.count`, and does
  not read more of the buffer than needed to return the desired
  number of lines.

- Only when asked to split a stream of just newlines, and into
  groups as large or larger than the chunk size, does the old code
  outperform the new by ~1.6x, but this corner case is not the one
  we should optimise for.

This supersedes #15, see there for rationale.

### CODE REVIEW

**Comment by fosskers** [comment]
Fantastic, I will get this merged tomorrow (Monday).

**Comment by vdukhovni** [comment]
> Fantastic, I will get this merged tomorrow (Monday).

Thanks. Please do take a careful look just in case I missed something.  I am confident it is correct, and ran a bunch of my own tests, but not to the point of discouraging review.

**Comment by fosskers** [comment]
I will, especially this since is a fairly core library in the `streaming` ecosystem.

File: Data/ByteString/Streaming/Char8.hs [hunk 1]
```diff
@@ -633,38 +633,52 @@ lineSplit :: forall m r. Monad m
...
+{-# INLINABLE lineSplit #-}
+
+-- | Return either how many newlines a strict bytestring chunk contains, if
+-- fewer than the number requested, or, else the total length of the requested
+-- number of lines within the bytestring (equivalently, i.e. the start index of
+-- the first /unwanted line/).
+nthNewLine :: B.ByteString   -- input chunk
+           -> Int            -- remaining number of newlines wanted
+           -> Either Int Int -- Left count, else Right length
+nthNewLine (B.PS fp off len) targetLines =
```
Comment (chessai) [inline Data/ByteString/Streaming/Char8.hs:667]: perhaps we should avoid using the `PS` constructor?

File: Data/ByteString/Streaming/Char8.hs [hunk 2]
```diff
@@ -633,38 +633,52 @@ lineSplit :: forall m r. Monad m
...
   -> Stream (ByteString m) m r
-lineSplit !n0 text0 = loop1 0 text0
+lineSplit !n0 text0 = loop1 text0
   where
     n :: Int
     !n = max n0 1
-    loop1 :: Int -> ByteString m r -> Stream (ByteString m) m r
-    loop1 !counter text =
+    loop1 :: ByteString m r -> Stream (ByteString m) m r
+    loop1 !text =
```
Comment (chessai) [inline Data/ByteString/Streaming/Char8.hs]:
```suggestion
    loop1 text =
```

This bang is redundant - it is immediately followed by `case text of { ... }`

**Comment by chessai** [comment]
@vdukhovni overall this looks great - left some minour comments.

File: Data/ByteString/Streaming/Char8.hs (hunk 1 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs:667]: I don't see how to do that, and stay performant.  The entire loop needs to happen under the control of a single `withForeignPtr`.  The PS constructor is already used elsewhere in this codebase.  It is not going away, just becomes a pattern-synonym in `bytestring` 0.11, but I've added a version bound for that in #16 .

File: Data/ByteString/Streaming/Char8.hs (hunk 2 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs]: Sure.  Will drop it.

File: Data/ByteString/Streaming/Char8.hs (hunk 1 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs:667]: What I could do is use `CPP` to determine whether `bytestring` is at least 0.11, and directly use the `BS` constructor eliminating `off`, but I hope that the compiler will do most of that for me via the pattern synonym and optimization.

So probably not worth the effort.

File: Data/ByteString/Streaming/Char8.hs (hunk 2 above)
Comment (vdukhovni) [reply Data/ByteString/Streaming/Char8.hs]: Pushed fixup.

**Comment by vdukhovni** [comment]
Just a gentle ping, when you have a chance...

**Comment by fosskers** [comment]
I will make a release tomorrow (Saturday). Thanks again for this!

**Pull request merged**